llvm/lib/Target/AIE/AIE2InstructionSelector.cpp

```diff
@@ -4044,92 +4179,120 @@ bool AIE2InstructionSelector::selectG_AIE_STORE_PACK(MachineInstr &StoreI,
-    Register ShftReg, Register SignReg, bool ConstantSign,
-    MachineRegisterInfo &MRI) {
-  assert(LSO.OffsetOpcode && "Expected an offset opcode for 512-bit VST.SRS!?");
-
+std::tuple<Register, Register>
+AIE2InstructionSelector::split512BitReg(Register SrcReg,
+                                        MachineRegisterInfo &MRI) {
   const TargetRegisterClass *RC = &AIE2::ACC512RegClass;
   Register Low512 = MRI.createVirtualRegister(RC);
   Register High512 = MRI.createVirtualRegister(RC);
+  auto LowerBits = MIB.buildInstr(TargetOpcode::COPY, {Low512}, {})
+                       .addReg(SrcReg, 0, AIE2::sub_512_lo);
+  auto HigherBits = MIB.buildInstr(TargetOpcode::COPY, {High512}, {})
+                        .addReg(SrcReg, 0, AIE2::sub_512_hi);
+  return {LowerBits.getReg(0), HigherBits.getReg(0)};
+}
+
+bool AIE2InstructionSelector::select512BitG_AIE_STORE_SRS(
+    StoreSRS512BitInfo SSI, AddressingModeInfo &AMI, Register SrcRegLo,
+    Register SrcRegHi, MachineRegisterInfo &MRI,
+    std::optional<StoreSRS512BitInfo> SSIHi) {
+  LoadStoreOpcodes LSOHi;
+  Register ShftRegHi, SignRegHi;
+  if (SSIHi) {
+    LSOHi = SSIHi->LSO;
+    ShftRegHi = SSIHi->ShftReg;
+    SignRegHi = SSIHi->SignReg;
+  } else {
+    LSOHi = SSI.LSO;
+    ShftRegHi = SSI.ShftReg;
+    SignRegHi = SSI.SignReg;
+  }
+  assert(LSOHi.OffsetOpcode &&
+         "Expected an offset opcode for 512-bit VST.SRS!?");
 
   switch (AMI.MemI.getOpcode()) {
   case AIE2::G_STORE:
   case AIE2::G_AIE_POSTINC_STORE:
   case AIE2::G_AIE_POSTINC_2D_STORE:
   case AIE2::G_AIE_POSTINC_3D_STORE: {
-    auto LowerBits = MIB.buildInstr(TargetOpcode::COPY, {Low512}, {})
-                         .addReg(SrcReg, 0, AIE2::sub_512_lo);
-    auto HigherBits = MIB.buildInstr(TargetOpcode::COPY, {High512}, {})
-                          .addReg(SrcReg, 0, AIE2::sub_512_hi);
-    auto StoreHigher = MIB.buildInstr(*LSO.OffsetOpcode)
+    auto StoreHigher = MIB.buildInstr(*LSOHi.OffsetOpcode)
                            .addReg(AMI.PtrOp.getReg())
                            .addImm(32) // Offset
-                           .addReg(HigherBits.getReg(0))
-                           .addReg(ShftReg);
+                           .addReg(SrcRegHi)
+                           .addReg(ShftRegHi);
 
-    auto StoreLower = MIB.buildInstr(LSO.ISelOpcode);
+    auto StoreLower = MIB.buildInstr(SSI.LSO.ISelOpcode);
 
     for (auto Def : AMI.MemI.defs())
       StoreLower.addDef(Def.getReg());
 
-    addAddressingMode(StoreLower, AMI, LSO.FitsImmediateRange, false, MRI);
+    addAddressingMode(StoreLower, AMI, SSI.LSO.FitsImmediateRange, false, MRI);
 
-    StoreLower.addReg(LowerBits.getReg(0)).addReg(ShftReg);
+    StoreLower.addReg(SrcRegLo).addReg(SSI.ShftReg);
 
     addSplitMemOperands(AMI.MemI, StoreHigher, StoreLower, 0, 2);
 
-    if (!ConstantSign)
-      setUnsetCtrlRegister(*StoreHigher, *StoreLower, MRI, AIE2::crSRSSign,
-                           SignReg);
+    if (SSIHi) {
```

Comment:
Also this part can be simplified.

Reply:
If we remove this logic, we have to use `setUnsetCtrlRegister` twice, which is adding extra instructions in existing tests.

Reply:
I think there is no problem, we already have logic to remove/hoist such duplicated moves.